### PR TITLE
[Fix #2719] App does not fit on iPhone X screen

### LINKED ISSUE #2719: Black bars are shown when using iPhone X
[comment]: # (Please replace ... with your information. Remove < and >)

### User Story

As a iPhone X user, I want to use Status so that it is optimized for my device.

### Description

[comment]: # (Feature or Bug? i.e Type: Bug)
*Type*: Bug

[comment]: # (Describe the feature you would like, or briefly summarise the bug and what you did, what you expected to happen, and what actually happens. Sections below)
*Summary*: There are black bars shown at the top and bottom of an iPhone X screen. As a user, this is jarring and detracts from the user experience. Instead, the app should take up the entire screen and be optimized for the safe areas and notch.

#### Expected behavior
[comment]: # (Describe what you expected to happen.)
Expected behavior is for the app to fill the black bars, respect safe areas, have the status bar display correctly, and cater to the notch.

#### Actual behavior
[comment]: # (Describe what actually happened.)
See the black bars in the screenshot.

img_0018 2

### Reproduction
[comment]: # (Describe how we can replicate the bug step by step.)
Open Status on an iPhone X.

### Solution
[comment]: # (Please summarise the solution and provide a task list on what needs to be fixed.)
*Summary*:  Apple has a full list of [required updates](https://developer.apple.com/ios/update-apps-for-iphone-x/).

Also possibly relevant, React Native introduced a [new API in 0.50.1](https://medium.com/react-native-training/react-native-iphonex-92ff511282af) for easily dealing with the safe area of an iPhone X. More [info here](https://blog.callstack.io/react-native-meets-iphonex-a0f5034f7a41).

### Additional Information
[comment]: # (Please do your best to fill this out.)
* Status version: 0.9.12
* Operating System:iOS

#### Logs

[comment]: # (if on Android please replicate bug whilst running adb logcat)
```
...
```

### PR DESCRIPTION
**Pull request by foopang**
fixes #2719 

I guess we will need another splash image for iPhone X portrait (1125px x 2436px) 🙂

screen shot 2017-12-30 at 9 42 17 pm

screen shot 2017-12-30 at 9 39 35 pm

status: ready

### CODE REVIEW

**Comment by denis-sharypin** [comment]
Hey, @foopang thanks for a contribution. Could you, please double check tab bar and top bar heights? Use my design as a reference. 
iphonex

**Comment by foopang** [comment]
Hi @denis-sharypin please see if this looks good now.

screen shot 2018-01-03 at 6 29 25 am

File: src/status_im/ui/screens/views.cljs [hunk 1]
```diff
@@ -113,9 +113,25 @@
...
+                                                        ;; iPhone X needs a background color to match the current view
+                                                        {:background-color
+                                                         (case current-view
+                                                           (:wallet
+                                                            :wallet-send-transaction
+                                                            :wallet-request-transaction
+                                                            :choose-recipient) common-styles/color-blue4
+                                                           (:accounts :login) common-styles/color-blue2
+                                                           :transparent)
+                                                         :margin-top -20}))
```
Comment (foopang) [inline src/status_im/ui/screens/views.cljs]: Seems that the new `safe-area-view` RN component doesn't allow us to set the padding size for top and bottom, so I come up with this kinda odd `:margin-top -20` to do the trick.

**Comment by denis-sharypin** [comment]
@foopang  Thanks! All good from a design side

**Comment by denis-sharypin** [comment]
@foopang Please, add also a launch screen 
launchscreen1125x2436

**Comment by foopang** [comment]
@denis-sharypin the launch screen has been updated!

screen shot 2018-01-03 at 8 37 05 pm

**Comment by oskarth** [comment]
@foopang Sorry, got dropped out of our pipeline. Moving to REVIEW now as design seems to have been sorted. Thanks!

File: src/status_im/ui/screens/views.cljs [hunk 2]
```diff
@@ -113,9 +113,25 @@
...
-                          (throw (str "Unknown view: " current-view)))]
+                          (throw (str "Unknown view: " current-view)))
+              iphone-x?   (and ios?
+                               (let [{:keys [width height]} (-> (.get dimensions "window")
+                                                                (js->clj :keywordize-keys true))]
+                                 ;; iPhone X dimensions = 375 x 812
+                                 (and (= width 375) (= height 812))))]
           [(if android? menu-context view) common-styles/flex
-           [view common-styles/flex
+           [(if iphone-x? safe-area-view view) (merge common-styles/flex
```
Comment (oskarth) [inline src/status_im/ui/screens/views.cljs]: This is a bit gnarly, can we break this out in a fn or so?

File: src/status_im/ui/screens/views.cljs [hunk 3]
```diff
@@ -1,8 +1,8 @@
 (ns status-im.ui.screens.views
   (:require-macros [status-im.utils.views :refer [defview letsubs]])
   (:require [re-frame.core :refer [dispatch]]
-            [status-im.utils.platform :refer [android?]]
-            [status-im.ui.components.react :refer [view modal]]
+            [status-im.utils.platform :refer [android? ios?]]
+            [status-im.ui.components.react :refer [view safe-area-view modal dimensions]]
```
Comment (oskarth) [inline src/status_im/ui/screens/views.cljs]: Can we turn these into react/foo etc? Generally preferable (think these are leftovers)

File: src/status_im/ui/components/react.cljs [hunk 4]
```diff
@@ -36,6 +36,7 @@
 (def geolocation (when (exists? js/window)
                    js/navigator.geolocation.))
 (def view (get-class "View"))
+(def safe-area-view (get-class "SafeAreaView"))
```
Comment (oskarth) [inline src/status_im/ui/components/react.cljs]: Is this always available? What happens if it is not?

**Review by oskarth — CHANGES_REQUESTED** [review]
Minor comments, in general looks good

**Comment by oskarth** [comment]
^ @jeluard @flexsurfer please review

File: src/status_im/ui/screens/views.cljs [hunk 5]
```diff
@@ -113,9 +113,25 @@
                           :paste-json-text paste-json-text
                           :add-rpc-url add-rpc-url
                           :network-details network-details
-                          (throw (str "Unknown view: " current-view)))]
+                          (throw (str "Unknown view: " current-view)))
+              iphone-x?   (and ios?
+                               (let [{:keys [width height]} (-> (.get dimensions "window")
+                                                                (js->clj :keywordize-keys true))]
```
Comment (jeluard) [inline src/status_im/ui/screens/views.cljs]: Sounds very error prone. Is that the official to do this?

**Review by jeluard — CHANGES_REQUESTED** [review]
Seems very hacky. Can you provide link to docs detailing official guidelines?

File: src/status_im/ui/components/react.cljs (hunk 4 above)
Comment (foopang) [reply src/status_im/ui/components/react.cljs]: Yes, it is always available. React Native ships with `SafeAreaView` as of version 0.50.

**Comment by foopang** [comment]
@jeluard I am not sure if there are official guidelines dealing with iPhone X from React Native docs. I followed this as a reference https://blog.callstack.io/react-native-meets-iphonex-a0f5034f7a41

File: src/status_im/ui/components/react.cljs [hunk 6]
```diff
@@ -202,3 +203,25 @@
       [view {:style style}
        [activity-indicator {:animating true}]]
       comp)))
+
+(defn create-main-screen-view [current-view]
+  (let [iphone-x?    (and ios?
+                          (let [{:keys [width height]} (-> (.get dimensions "window")
```
Comment (jeluard) [inline src/status_im/ui/components/react.cljs]: Can we extract this logic somewhere?

File: src/status_im/ui/components/react.cljs [hunk 7]
```diff
@@ -202,3 +203,25 @@
...
+                                                           (js->clj :keywordize-keys true))]
+                            ;; iPhone X dimensions = 375 x 812
+                            (and (= width 375) (= height 812))))
```
Comment (jeluard) [inline src/status_im/ui/components/react.cljs]: Please use constant for magic number

**Review by jeluard — CHANGES_REQUESTED** [review]

**Comment by foopang** [comment]
@jeluard updated, please have a look.

**Review by jeluard — APPROVED** [review]

**Comment by jeluard** [comment]
@foopang Thanks!

**Review by oskarth — APPROVED** [review]

**Comment by Serhy** [comment]
@foopang there is conflict appeared, could you rebase with develop please? 

Additionally, we tested on iPhoneX (iOS 11.2.2) the build just right before the conflict -  https://i.diawi.com/QNR1AK (0.0.2157 version).

Adjusting the views to iPhoneX we introduce several issues we need to consider before merge:

**1)** On QR code screens (both - wallet and public key). Could you please move the button with the key value and QR code a bit higher so the button not overlapped by home indicator.
screen shot 2018-01-25 at 18 28 18


**2)** The **Contacts** view (_Wallet -> Send transaction -> Choose recipient… -> Choose from contacts_) and **Transaction Fee** view headers (_Wallet -> Send transaction -> Advanced -> Transaction Fee_) are too close to device status icons menu. Could you move both screens a bit lower please? (Green line on Transaction Fee screen just to show the edge of the device status icons menu)
screen shot 2018-01-25 at 18 31 47
screen shot 2018-01-25 at 18 32 10

**3)** There is a gap between keyboard and message editbox appears when in 1-1/group/public chat user types a message. Could you move please messagebox area lower and join it with the keyboard?
screen shot 2018-01-25 at 18 33 01

**4)** Pop-up screen menu with the list of contacts from /request option in Group Chat overlapped by device status icons menu. (to reproduce - in any Group Chat select /request option and pull-top the appeared menu with the list of contacts)
screen shot 2018-01-25 at 18 36 53

5) The @browse screen extended pretty much so it’s overlapped by device status icons menu. **To reproduce**: Open Status -> Create new account -> create password -> in Console tap on @browse -> type amazon.com -> tap Send button -> **Result**: the first screenshot below -> select 'Switch to fullscreen' at the bottom-right -> **Result**: the second screenshot below). 
screen shot 2018-01-25 at 18 41 50
screen shot 2018-01-25 at 18 42 08

6) The **Assets** (_Wallet -> top-right menu button -> Manage Assets_) and **Filter history** (_Wallet -> Transaction history -> Top right menu button_) screens overlapped by device status icons menu. It is not possible to tap on ‘Done’ button to return to previous screen. 
screen shot 2018-01-25 at 18 44 02
screen shot 2018-01-25 at 18 44 40

7) The tab bar on the Wallet main screen is cut off, - not the same as on Profile or Home screens. Could you extend please the white background to the bottom of the tab bar on Wallet screen view.
screen shot 2018-01-25 at 18 49 46

**Comment by foopang** [comment]
I’ve refactored the code a bit in order to address the issues found by @Serhy 

For issue #5 reported, it is not always reproducible. I suspect that is a bug of `react-native-webview-bridge` but I couldn’t fix it yet.

screen shot 2018-01-27 at 2 00 28 pm
screen shot 2018-01-27 at 2 01 08 pm
screen shot 2018-01-27 at 2 08 40 pm
screen shot 2018-01-27 at 2 47 36 pm
screen shot 2018-01-27 at 4 02 11 pm
screen shot 2018-01-27 at 7 26 00 pm
screen shot 2018-01-27 at 8 00 04 pm
screen shot 2018-01-28 at 5 31 07 pm

**Comment by jeluard** [comment]
@foopang Have you considered handling this at the `status-bar` level instead of relying on `platform-specific-view`?

**Comment by foopang** [comment]
@jeluard Oh, you're right! I can simplify it a little bit and also remove the odd looking `margin-top -20` in `platform-specific-view` by just setting `:status-bar-default-height` to 0 for iPhone X. Thanks for the useful tip!

**Comment by foopang** [comment]
@Serhy I think the @browse screen issue is not relevant anymore since we removed the @browse command and implemented our own browser. Can you check this again? 

screen shot 2018-02-04 at 11 22 35 am

**Comment by Serhy** [comment]
Thanks a lot @foopang 
Build for iOS: https://i.diawi.com/oUg64n has been tested on iPhoneX.

Yes, all above issues are fixed, and with the removed @browse screen - 5th issue from https://github.com/status-im/status-react/pull/2856#issuecomment-360741904 is not relevant anymore. A couple of things appeared we need to look at (I could not reproduce them in latest develop which is 0.9.11d632):

On `Wallet` -> `Send transaction` and `Wallet` -> `Receive transaction` screens there is bottom bar with buttons: 'Sign later'/'Sign transaction' (for 'Send transaction' screen) and 'Share'/'Send request' (for 'Receive transaction' screen). 
These bars apparently behind keyboard, when the last is present on screen, so there is no way to proceed with the request further until keyboard is hidden. Please see steps to reproduce what I mean:

1) Open Status and log in to account
2) Navigate to Wallet -> 'Send transaction'
3) 'Choose recipient' -> 'Choose from contacts' -> tap 'Jarrad'
4) Tap in 'Specify amount' 
5) Tap 'Sign later'/'Sign transaction'

**Actual result(1)**: 
image uploaded from ios 2

6) Navigate back to Wallet main screen
7) Back to 'Send transaction' 
8) Tap Specify amount and type 0
9) Tap 'Choose recipient' -> Choose From Contacts -> Jarrad (Note: in such case the bottom bar with 'Sign later' and 'Sign transaction' buttons are visible, however:
10) Tap 'Sign transaction' enter password and look if 'Cancel' with 'Sign transaction' buttons are present

**Actual result(2)**: 
image uploaded from ios 5


11) Navigate to 'Wallet' -> 'Receive transaction' screen 
12) Tap 'Specify amount' and type 0
13) Tap 'Send request' 

**Actual request(3)**: 
image uploaded from ios 3

And one more issue with 'Sign transaction' screen
1) From 'Home' screen -> tap `+` (Start new chat) -> Jarrad (to start 1-1 chat)
2) Tap /send command -> enter 0 value -> 'Send message' blue button
3) Once you wind up on 'Send transaction' screen check bottom bar with the 'Sign later' and 'Sign transaction' buttons

**Actual result(4)**: 
image uploaded from ios 4


Could you move them upper a bit so it would not be overlapped by the device home indicator?

**Comment by janherich** [comment]
@foopang Please resolve conflicts

**Comment by foopang** [comment]
@Serhy The transaction screen issues have been fixed now. Please test again and let me know if there are any issues still. Thanks!

screen shot 2018-02-06 at 11 12 55 am
screen shot 2018-02-06 at 11 17 40 am
screen shot 2018-02-06 at 11 19 18 am
screen shot 2018-02-06 at 3 43 27 pm

**Comment by Serhy** [comment]
Hey @foopang, thanks! All above fixed in latest build for iOS: https://i.diawi.com/V2TuaJ.
But there is one regression appeared (might be related to fix of the last one issue where transaction is send and processed from chat):

There is no 'Transaction sent' confirmation screen after transaction sign in, if Tx was created from 1-1/group chats or signed from 'Transactions' -> 'Unsigned' screen. App is no responsive for tap on any part of screen until re-open the app. 
Please see steps to reproduce: 

**Preconditions**: Account is created and ETH balance > 0.0001 (/faucet request sent) 

1) Open Status and log in to account
2) Navigate to 'Home' -> 'Start new chat' -> 'Jarrad'
3) Tap on /send option -> type valid amount to send (e.g. 0.00003) -> 'Send' button
4) From 'Send transaction' ->tap  'Sign transaction' button -> enter valid password -> 'Sign transaction' button
5) Check if you wind up on 'Transaction sent' screen

Actual result: Redirect to 1-1 chat screen. App is not responsive. 
image uploaded from ios 7


6) Re-open Status application and log in with the same account
7) Navigate to 'Wallet' -> 'Send transaction'
8) Specify recipient -> Recent recipients -> Jarrad
9) Specify amount -> type valid amount (e.g. 0.00005)
10) Tap 'Sign later' -> Yes
11) Navigate to 'Transaction history' -> 'Unsigned' tab -> Tap 'Sign' button on unsigned transaction
12) From 'Send transaction' ->tap  'Sign transaction' button -> enter valid password -> 'Sign transaction' button 
13) Check if you wind up on 'Transaction sent' screen

Actual result: Redirect to 'Transactions' screen. App is not responsive. 
image uploaded from ios 6

Note: on both screens above - there is not visible device status bar at the very top.


And one more thing I have to note (after rebase from develop appeared):
The camera view screen which accessed from 'Wallet' -> 'Send transaction' -> 'Scan QR code': the there is a rectangle instead of square for QR scanning area and two black unexpected squares below are present at the moment:

screen shot 2018-02-06 at 17 36 20

**Comment by foopang** [comment]
@Serhy Sorry for bringing in so many issues in this PR. Hope all is fixed now! Please have a look.

**Comment by Serhy** [comment]
Thanks a lot, @foopang !
We checked https://i.diawi.com/nZrTp8 build on iPhoneX and iPhone 5s (and Android 7.0 too!) as well this PR and it was noticed small regression:
On **Assets** screen - the device status bar with white background while it was a blue one.

**iPhoneX view:**
image uploaded from ios 11

**iPhone 5s view**:


**Develop and mockup view**:
screen shot 2018-02-08 at 16 49 03


Could you please apply on 'Wallet' -> top-right button -> 'Manage Assets' -> 'Assets' screen the blue background to device status bar and fix conflicts?

**Comment by flexsurfer** [comment]
@foopang could you please resolve conflicts and take a look on @Serhy comment, thanks

**Review by flexsurfer — APPROVED** [review]

**Comment by foopang** [comment]
@Serhy Thanks for checking! All layout issues should have been fixed now.

@flexsurfer @jeluard  something to note about the iPhone X screen issues, I did some hacks to workaround.

https://github.com/status-im/status-react/pull/2856/files#diff-d8aa2157cacb46da51afecaef8554e32R294 The bottom spacing is gone when in modal view. So I put an invisible `view` element to the bottom for adding bottom spacing. but it is not needed when the keyboard is opened up for signing transaction.

https://github.com/status-im/status-react/pull/2856/files#diff-93d9c10db297821b5145f29bdbb7b6bcR189 The transaction sent modal won't show but will block the layout from being navigated away when closing the transaction modal and immediately opening a new modal, so adding a delay here to fix.

https://github.com/status-im/status-react/pull/2856/files#diff-dc0057ed633d7dc9f584171ce43774d5R28 The top and bottom spacing has to be reduced from the height in order to get the screen viewport height of iPhone X so that "Scan QR code" screen can show properly.

https://github.com/status-im/status-react/pull/2856/files#diff-d8aa2157cacb46da51afecaef8554e32R275 The top bar and the bottom bar can't be two different background colors. To work around that I added a view with a different background color and stuck to the bottom so that it can show different background colors on Wallet screens.


<blockquote><div> GitHub</div><div><strong><a href="https://github.com/status-im/status-react/pull/2856">[Fix #2719] App does not fit on iPhone X screen by foopang · Pull Request #2856 · status-im/status-react</a></strong></div><div>fixes #2719
I guess we will need another splash image for iPhone X portrait (1125px x 2436px) 🙂


status: ready</div></blockquote>
<blockquote><div> GitHub</div><div><strong><a href="https://github.com/status-im/status-react/pull/2856">[Fix #2719] App does not fit on iPhone X screen by foopang · Pull Request #2856 · status-im/status-react</a></strong></div><div>fixes #2719
I guess we will need another splash image for iPhone X portrait (1125px x 2436px) 🙂


status: ready</div></blockquote>
<blockquote><div> GitHub</div><div><strong><a href="https://github.com/status-im/status-react/pull/2856">[Fix #2719] App does not fit on iPhone X screen by foopang · Pull Request #2856 · status-im/status-react</a></strong></div><div>fixes #2719
I guess we will need another splash image for iPhone X portrait (1125px x 2436px) 🙂


status: ready</div></blockquote>
<blockquote><div> GitHub</div><div><strong><a href="https://github.com/status-im/status-react/pull/2856">[Fix #2719] App does not fit on iPhone X screen by foopang · Pull Request #2856 · status-im/status-react</a></strong></div><div>fixes #2719
I guess we will need another splash image for iPhone X portrait (1125px x 2436px) 🙂


status: ready</div></blockquote>

**Comment by Serhy** [comment]
Builds Android: https://i.diawi.com/jH87hH and iOS: https://i.diawi.com/ND1hGn has been tested on iPhoneX and Android 7.0 (Samsung J7).

It looks good, no issues and app fits on iPhoneX screen.

@foopang great job! Thank you a lot for the efforts in this PR!

Ready for merge!

**Comment by chadyj** [comment]
Thank you @foopang! This has been a much requested feature and will be a fantastic enhancement to Status.

**Pull request merged**